Add support for resolving $(MSBuildExtensionsPath) with fallback sear…

radical opened the pull request:
…ch paths.

In Mono's xbuild, $(MSBuildExtensionsPath) is $mono_prefix/lib/mono/xbuild.
On OS X, this is:

    `/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/xbuild`

Consequently, when resolving an Import for
$(MSBuildExtensionsPath)\Foo\Abc.targets, the following path will be checked:

    `/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/xbuild/Foo/Abc.targets`

This works, but has a problem when upgrading the Mono.framework package:
a *new* /Library/Frameworks/Mono.framework/Versions/@VERSION@ directory
is created, so anything installed into the previous directory will be
"lost". The file is still there, but xbuild will no longer find it.

To address this upgrade scenario, Mono's xbuild also checks a fallback path:

    `/Library/Frameworks/Mono.framework/External/xbuild`

This location is not within a versioned framework directory, and thus files
installed into it will survive Mono.framework package upgrades.

Unfortunately, this means that $(MSBuildExtensionsPath) on xbuild is no longer
a single directory; it is a set of directories, and xbuild will check each
directory to see if the file exists:

/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/xbuild/Foo/Abc.targets
/Library/Frameworks/Mono.framework/External/xbuild/Foo/Abc.targets
...

Mono would like to eventually migrate to MSBuild instead of xbuild, which
requires bringing xbuild's fallback support to MSBuild.

These search paths can now be specified in the app.config like:

      <toolset toolsVersion="14.1">

        <msbuildExtensionsPathSearchPaths>
            <searchPaths os="osx">
                <property name="MSBuildExtensionsPath" value="/Library/Frameworks/Mono.framework/External/xbuild/;/tmp/foo"/>
                <property name="MSBuildExtensionsPath32" value="/Library/Frameworks/Mono.framework/External/xbuild/"/>
                <property name="MSBuildExtensionsPath64" value="/Library/Frameworks/Mono.framework/External/xbuild/"/>
            </searchPaths>
        </msbuildExtensionsPathSearchPaths>

      </toolset>

AndyGerlicher commented:
Just to give a heads up we do want this (perhaps even in master). We just need time to review and make sure it's an approach we're comfortable with and give feedback.

radical commented:
Thanks for the heads up! I can wait :)

AndyGerlicher commented:
Overall, here are our comments. Hopefully you're able to address them and we can get this committed. 

- Remove the feature flag. We'd like to move this to master and if it's going to be enabled in all our configs might as well remove it. 
- Few minor code changes (I'll comment on them individually)
- We need a better way to implement search paths and not hard code anything environment specific. Our proposal:
  - Use / read from `MSBuildExtensionsPathSearchPaths{OS}`
  - Move the `MSBuildExtensionsPathSearchPaths{OS}` property to the [toolsets section in the App.config](https://github.com/Microsoft/msbuild/blob/master/src/XMakeCommandLine/app.config#L38-L40).
  - This would allow us to have `MSBuildExtensionsPathSearchPathsOSX` set to `/Library/Frameworks/Mono.framework/External/xbuild` and `MSBuildExtensionsPathSearchPathsWindows` set to `C:\Program Files (x86)\MSBuild` (see below)

And of course feel free to propose something else if you can think of something better. Our main goal is we want to be able to take this in master in a generic way. We had discussed doing something like this as helping our OSS version work when it doesn't live in `Program Files` but other things do.

Thanks!

AndyGerlicher commented on src/XMakeBuildEngine/Evaluation/MSBuildExtensionsPathHelper.cs:
```diff
@@ -0,0 +1,164 @@
+            }
+        }
+
+        private static List<string> ComputeFallbackExtensionsPathsFor(MSBuildExtensionsPathReferenceKind refKind)
```
This method isn't called

AndyGerlicher commented on src/XMakeBuildEngine/Evaluation/MSBuildExtensionsPathHelper.cs:
```diff
@@ -0,0 +1,164 @@
+
+            var allPaths = String.Join(Path.PathSeparator.ToString(),
+                                        //FIXME: IsOSX returns false :/
+                                        NativeMethodsShared.IsUnix ? s_osxExternalXBuildDir : String.Empty,
```
Did you mean IsUnixLike? Either way with the solution I proposed hopefully this should go away and be completely generic here.

AndyGerlicher commented on src/XMakeBuildEngine/Evaluation/MSBuildExtensionsPathHelper.cs:
```diff
@@ -0,0 +1,164 @@
+                                            envVar ?? String.Empty);
+
+                var pathsTable = new Dictionary<string, string>();
+                foreach(var extnPath in allPaths.Split(new char[]{Path.PathSeparator}, StringSplitOptions.RemoveEmptyEntries)) {
```
Style: { on a new line and space after foreach and if. Also we used to run StyleCop which would also require a {} around every if. It's not a hard requirement, but this would be the only place not conforming so you might as well. The CodeFormatter tool is great for auto correcting these style things, but we haven't ran that in a while it might have a lot of hits.

AndyGerlicher commented on src/XMakeBuildEngine/Evaluation/MSBuildExtensionsPathHelper.cs:
```diff
@@ -0,0 +1,164 @@
+            s_pathsLists = new Lazy<ExtensionsPathsLists>(true);
+        }
+
+        class ExtensionsPathsLists
```
Please be explicit on visibility of types / members (in the test class too).

AndyGerlicher commented on src/XMakeBuildEngine/Evaluation/MSBuildExtensionsPathHelper.cs:
```diff
@@ -0,0 +1,164 @@
+        }
+    }
+
+    enum MSBuildExtensionsPathReferenceKind {
```
Please add /// doc and explicit `internal`

cdmihai replied:
You could convert this enum into a struct with private constructor and four static values. Then, you can move all the enum specific static methods in the struct. With this you can also convert the ReferenceKindToString method into a an extra field of the struct (like StringRepresentation).

cdmihai commented on src/XMakeBuildEngine/Evaluation/MSBuildExtensionsPathHelper.cs:
```diff
@@ -0,0 +1,164 @@
+                    if(!Directory.Exists(extnPath))
+                        continue;
+
+                    pathsTable[extnPath] = extnPath;
```
Looks a bit weird to have the key and value be the same thing. If you want to check for unique entries, can you do it with a HashSet instead, and thus also skip the first if in the loop?

cdmihai commented on src/XMakeBuildEngine/Evaluation/MSBuildExtensionsPathHelper.cs:
```diff
@@ -0,0 +1,164 @@
+                    if(pathsTable.ContainsKey(extnPath))
+                        continue;
+
+                    if(!Directory.Exists(extnPath))
```
Why not let callers check for existence? What if the caller code wants to do some funky lazy creation of the paths?

cdmihai commented on src/XMakeBuildEngine/Evaluation/MSBuildExtensionsPathHelper.cs:
```diff
@@ -0,0 +1,164 @@
+            /// <summary>
+            /// Cached list of extension paths to be used as search paths for $(MSBuildExtensionsPath)
+            /// </summary>
+            public readonly List<string> Default;
```
You could convert these 3 fields into a single Dictionary<MSBuildExtensionsPathReferenceKind, List<string>> and have GetFallbackExtensionsPathsFor delegate here instead of having that switch statement

cdmihai commented on src/XMakeBuildEngine/Evaluation/Evaluator.cs:
```diff
@@ -2062,6 +2206,7 @@ private List<ProjectRootElement> ExpandAndLoadImports(string directoryOfImportin
 
             foreach (string importFileEscaped in importFilesEscaped)
             {
+                atleastOneImportProcessed = true;
```
Can you move this to line 2324, right after it gets added in the imports list? Then, it really means that at least one got processed. Otherwise rename boolean flag to atLeastOneFound.

cdmihai commented on src/XMakeBuildEngine/Evaluation/Evaluator.cs:
```diff
@@ -2026,15 +2021,164 @@ private bool EvaluateWhenOrOtherwiseChildren(IEnumerable<ProjectElement> childre
+            string importExpressionEscaped = _expander.ExpandIntoStringLeaveEscaped(unescapedExpression, ExpanderOptions.ExpandProperties, importElement.ProjectLocation);
             ElementLocation importLocationInProject = importElement.Location;
 
+            atleastOneImportFoundAndIgnored = false;
```
Rename to atleastOneImportIgnored.

cdmihai commented on src/XMakeBuildEngine/Evaluation/Evaluator.cs:
```diff
@@ -2026,15 +2021,164 @@ private bool EvaluateWhenOrOtherwiseChildren(IEnumerable<ProjectElement> childre
+            // Making a local copy for this instance of Evaluator, to add the the value of
+            // $(MSBuildExtensionsPath*) property that we have here, instead of keeping it
+            // in the static/shared MSBuildExtensionsPathHelper
+            var allExtensionPaths = new List<string>(fallbackExtensionPaths);
```
Make the helper be proactive and clone its lists before returning.

cdmihai commented on src/XMakeBuildEngine/Evaluation/Evaluator.cs:
```diff
@@ -2026,15 +2021,164 @@ private bool EvaluateWhenOrOtherwiseChildren(IEnumerable<ProjectElement> childre
+            var allExtensionPaths = new List<string>(fallbackExtensionPaths);
+            allExtensionPaths.Insert(0, _data.GetProperty(MSBuildExtensionsPathHelper.ReferenceKindToString(refKindInProject)).EvaluatedValue);
+
+            string extensionPropertyRefAsString = String.Format("$({0})", MSBuildExtensionsPathHelper.ReferenceKindToString(refKindInProject));
```
Move this to the future struct for refKind. Something like refKindInProject.ToMSBuildPropertyName()

cdmihai commented on src/XMakeBuildEngine/Evaluation/Evaluator.cs:
```diff
@@ -2026,15 +2021,164 @@ private bool EvaluateWhenOrOtherwiseChildren(IEnumerable<ProjectElement> childre
+                }
+
+                // Nothing was ignored, but if @atleastOneImportProcessed==true, then we did try
+                // to look at a file path and found it missing!
```
Why the conclusion that it was missing? atleastOneImportProcessed is set to true right at the beginning of the loop, and it may also mean that that path was found and processed into a ProjectRootElement

radical replied:
I have tried to make the code a little clearer in the updated patch.

radical commented:
@AndyGerlicher Thanks for the feedback! `app.config` approach sounds good. Do you want the properties to be per toolset? Like:

      <toolset toolsVersion="14.1">   
        <property name="MSBuildToolsPath" value="." />   

        <property name="MSBuildExtensionsPathSearchPathsWindows" value="...." />
        <property name="MSBuildExtensionsPath32SearchPathsWindows" value="...." />
        <property name="MSBuildExtensionsPath64SearchPathsWindows" value="...." />

        <property name="MSBuildExtensionsPathSearchPathsOSX" value="...." />
        <property name="MSBuildExtensionsPath32SearchPathsOSX" value="...." />
        <property name="MSBuildExtensionsPath64SearchPathsOSX" value="...." />
      </toolset>

radical commented:
Or:  

      <toolset toolsVersion="14.1">   

         <msbuildExtensionsPathSearchPaths os="windows">
            <property name="MSBuildExtensionsPath" value="...." />
            <property name="MSBuildExtensionsPath32" value="...." />
            <property name="MSBuildExtensionsPath64" value="...." />
         </extensionPaths>

         <msbuildExtensionsPathSearchPaths os="osx">
            <property name="MSBuildExtensionsPath" value="...." />
            <property name="MSBuildExtensionsPath32" value="...." />
            <property name="MSBuildExtensionsPath64" value="...." />
         </extensionPaths>
      </toolset>

AndyGerlicher commented:
The second one looks better, but it might be overly complicated to implement. I would say go for the first:
```XML
<toolset toolsVersion="14.1">   
    <property name="MSBuildToolsPath" value="." />   

    <property name="MSBuildExtensionsPathSearchPathsWindows" value="...." />
    <property name="MSBuildExtensionsPath32SearchPathsWindows" value="...." />
    <property name="MSBuildExtensionsPath64SearchPathsWindows" value="...." />
```

radical commented:
I already implemented this:


    <msbuildToolsets default="14.1">
      <toolset toolsVersion="14.1">
        <property name="MSBuildToolsPath" value="." />
        <allMSBuildExtensionsPathSearchPaths>
            <msbuildExtensionsPathSearchPaths os="osx">
                <property name="MSBuildExtensionsPath" value="/Library/Frameworks/Mono.framework/External/xbuild/;/tmp/testing"/>
                <property name="MSBuildExtensionsPath32" value="/Library/Frameworks/Mono.framework/External/xbuild/"/>
                <property name="MSBuildExtensionsPath64" value="/Library/Frameworks/Mono.framework/External/xbuild/"/>
            </msbuildExtensionsPathSearchPaths>
            <msbuildExtensionsPathSearchPaths os="windows">
                <property name="MSBuildExtensionsPath" value="C:\\foo\\bar"/>
                <property name="MSBuildExtensionsPath64" value="C:\\foo\\bar"/>
            </msbuildExtensionsPathSearchPaths>
        </allMSBuildExtensionsPathSearchPaths>
      </toolset>
    </msbuildToolsets>

Open to suggestions for the new element names.
FYI, I added this to be part of the ToolsetElement and Toolset class. And added the "read" part of
it to ToolsetConfigurationReader. It seemed like a more logical place for this.

I'm just fixing tests now and will try to update this PR tomorrow for review.

AndyGerlicher commented:
Sounds good, thanks!

radical commented:
@AndyGerlicher @cdmihai Thanks for the detailed feedback! I have tried to incorporate your suggestions in the updated PR. I have split the changes into logical self contained commits.

radical commented:
Should the value of OS specified in ` <searchPaths os="osx">` be case insensitive? It isn't currently.

radical commented:
BTW, there is a fix required for IsOSX on mono, before this will work on OSX. I will open a new PR for that. But this can reviewed without that.

radical commented:
Updated to show error location in the config file.

radical commented:
BTW, PR #518 should be merged before this, else the tests will fail on OSX.

radical commented:
Any feedback on this?

AndyGerlicher commented:
Sorry to take a while to get back! This is a decently big feature I wanted to take the time to try it out in a few scenarios. Here some feedback.

1. If you import `$(MSBuildExtensionsPath)\*.props`, it will only import from the first folder that has a hit. I thought that in your previous iteration it imported the union of all props in all search paths? Did this change? Or am I imagining things? My first instinct says it should be the union of all the results that match that are distinct file names, but I'm not sure that's right. Thoughts on that?

2. The error message could be better. 

Previously:
```
D:\test\fallback.proj(3,2): error MSB4019: The imported project "C:\Program Files (x86)\MSBuild\14.0\notfound.props" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk.
```
Now:
```
D:\test\fallback.proj(3,10): error MSB4019: The imported project "$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\notfound.props" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk.
```

It really should note the normal expanded path, and ideally where else it looked. Debugging could end up being really tricky if not done.

Other than that looks good. If you can address # 2 I'd like to get this in. Thanks!

radical commented:
No problem. Thanks for taking the time and effort to review and give feedback :)

re:1, No, it stops at the first hit by design. Even the previous iteration had the same behavior. If you noticed otherwise, then that is a bug.

Thinking about the union behavior, it sounds like it might be useful for the `*.props` case, but ..

a. Security issue? Being able to inject targets, modify the build in subtle ways.

b. It would create dependence on the "environment", like random props file getting imported unintentionally. Wildcards might "catch"/import too much. It might be difficult to reason about stuff without being on the build machine.

IMHO, I don't think we should go for this.

re:2, error message. The details can actually be seen in a `/v:diag` log, like:

```
Search paths being used for $(MSBuildExtensionsPath) are /Users/ankit/dev/misc/lib/mono/xbuild:/Library/Frameworks/Mono.framework/External/xbuild/
Trying to import /Users/ankit/dev/misc/lib/mono/xbuild/non-existant.proj using extensions path /Users/ankit/dev/misc/lib/mono/xbuild
Trying to import /Library/Frameworks/Mono.framework/External/xbuild//non-existant.proj using extensions path /Library/Frameworks/Mono.framework/External/xbuild/
```

But this assumes that the user is aware that `$(MSBuildExtensionsPath)` is backed by multiple paths and they can see this with `/v:diag` , which might not be the case. And if the user is unaware, then s/he is lost! How about something like:

```
  /Users/ankit/test/msbuild/foo.proj(5,10): error MSB4019: The imported project "$(MSBuildExtensionsPath)/Bar/non-existant.proj" was not found. Tried to find the project file path `Bar/non-existant.proj` under the search paths - /Users/ankit/dev/misc/lib/mono/xbuild and /Library/Frameworks/Mono.framework/External/xbuild/ .  Confirm that the path in the <Import> declaration is correct, and that the file exists on disk in one of the search paths.
```

Maybe also mention that these are set in `/path/to/MSBuild.exe.config` ?

radical commented:
I think you were suggesting the error message to be like:
```
/Users/ankit/test/msbuild/foo.proj(5,10): error MSB4019: The imported project "/Users/ankit/dev/misc/lib/mono/xbuild/Bar/non-existant.proj" was not found. Also tried to find the project file path `Bar/non-existant.proj` under the search paths - /Library/Frameworks/Mono.framework/External/xbuild/ .  Confirm that the path in the <Import> declaration is correct, and that the file exists on disk in one of the search paths.
```

This shows the normal expanded path, expanded with the value of the property `$(MSBuildExtensionsPath)`. Which one is preferable?

rainersigwald commented:
@radical I think I like the form with the explicit default path, since that's most like the existing error. I might go a bit further though:

```
/Users/ankit/test/msbuild/foo.proj(5,10): error MSB4019: The imported project "/Users/ankit/dev/misc/lib/mono/xbuild/Bar/non-existant.proj" was not found. Since it was defined using $(MSBuildExtensionsPath), `Bar/non-existant.proj` was also seareched for in - /Library/Frameworks/Mono.framework/External/xbuild/ . Those paths are defined in MSBuild.exe.config.  Confirm that the path in the <Import> declaration is correct, and that the file exists on disk in one of the search paths.
```

The important addition being `Since it was defined using $(MSBuildExtensionsPath),`.

That's awfully wordy though.

radical commented:
> The important addition being Since it was defined using $(MSBuildExtensionsPath),.

Cool! I was thinking of mentioning that too, but I thought maybe I was being too cautious :)

How about - `... not found. Also, tried to find 'Bar/non-existant.proj' in the fallback search paths for $(MSBuildExtensionsPath) - /foo/bar and /xyz/abc . These search paths are defined in '/path/to/MSBuild.exe.config' . Confirm ...` ?

The *fallback* bit immediately suggests to the user the *order* of the search also. And it's consistent with the `Also, tried to..` .
I think the user can connect that `$(MSBuildExtensionsPath)` was relevant here because the import expression references it.

AndyGerlicher commented:
Sounds good @radical, not quite as wordy and conveys the search order I like that. And small subtlety since a lot of these paths are like `$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\...`, make sure that the rest of the path is expanded so the error doesn't have `$(MSBuildToolsVersion)` in it.

radical commented:
For an import `<Import Project="$(MSBuildExtensionsPath32)\$(Version)\non-existant.proj"/>`, the error message is:

  /Users/ankit/test/msbuild/foo.proj(8,10): error MSB4226: The imported project "/Users/ankit/dev/misc/lib/mono/xbuild/v0.1/non-existant.proj" was not found. Also, tried to find "v0.1/non-existant.proj" in the fallback search path(s) for $(MSBuildExtensionsPath32) - "/Library/Frameworks/Mono.framework/External/xbuild/" . These search paths are defined in "/Users/ankit/dev/msbuild/bin/Debug-MONO/OSX_Deployment/MSBuild.exe.config". Confirm that the path in the <Import> declaration is correct, and that the file exists on disk in one of the search paths.

I have not removed the diagnostic messages:

```
Search paths being used for $(MSBuildExtensionsPath32) are /Users/ankit/dev/misc/lib/mono/xbuild:/Library/Frameworks/Mono.framework/External/xbuild/
Trying to import /Users/ankit/dev/misc/lib/mono/xbuild/$(Version)/non-existant.proj using extensions path /Users/ankit/dev/misc/lib/mono/xbuild
Trying to import /Library/Frameworks/Mono.framework/External/xbuild//$(Version)/non-existant.proj using extensions path /Library/Frameworks/Mono.framework/External/xbuild/
```

Should these be removed now? I thought they might be useful in debugging, if say there is a crash before that error message shows up.

Another thing to note: I am using `AppDomain.CurrentDomain.SetupInformation.ConfigurationFile` to get the config file path, but I am not sure whether this will always be correct for our case.

radical commented:
Oh, .. `AppDomain` won't be in coreclr! I'll fix.

radical commented:
Another thing to note is that if no extension paths are found, then the error shown is the existing one. That is because in that case the code just executes the existing (regular import) code path.

And for CoreCLR case, the error message is missing the `These search paths ... .config .` part.

Build should be fixed now.

AndyGerlicher commented:
LGTM

rainersigwald commented on src/XMakeBuildEngine/UnitTests/Definition/ToolsetConfigurationReader_Tests.cs:
```diff
@@ -488,6 +500,182 @@ public void PropertyTest_GetValueByName()
+            ToolsetConfigurationSection msbuildToolsetSection = config.GetSection(s_msbuildToolsets) as ToolsetConfigurationSection;
+
+            Assert.Equal(msbuildToolsetSection.Default, "2.0");
+            Assert.Equal(1, msbuildToolsetSection.Toolsets.Count);
```
I was going to complain that the order of arguments is expected, actual--but this whole file is wrong. So leave things like this and maybe we can come along and fix it in the future.

radical replied:
Oops :( Sure, I can fix it after this is merged. Was there anything else wrong with the file?

radical replied:
Oh, the existing stuff is wrong too :/

rainersigwald replied:
Exactly . . . no point worrying about it now.

rainersigwald replied:
(and better to be consistent with the rest of the file than correct, in this case)

rainersigwald commented:
:+1:

radical commented:
Yay! Thanks for taking the time with the reviews and feedback :)
